encoding/custom/sema_codec/encode.go

```diff
@@ -0,0 +1,1101 @@
+	err = e.EncodeLocationPrefix(common.StringLocationPrefix)
+	if err != nil {
+		return
+	}
+
+	return e.EncodeString(string(t))
+}
+
+func (e *SemaEncoder) EncodeTransactionLocation(t common.TransactionLocation) (err error) {
+	err = e.EncodeLocationPrefix(common.TransactionLocationPrefix)
+	if err != nil {
+		return
+	}
+
+	return e.write(t[:])
+}
+
+func (e *SemaEncoder) EncodeREPLLocation() (err error) {
+	return e.EncodeLocationPrefix(common.REPLLocationPrefix)
+}
+
+// EncodeString encodes a string as a byte array.
+func (e *SemaEncoder) EncodeString(s string) (err error) {
+	return e.EncodeBytes([]byte(s))
+}
+
+// EncodeBytes encodes a byte array.
+func (e *SemaEncoder) EncodeBytes(bytes []byte) (err error) {
+	err = e.EncodeLength(len(bytes))
+	if err != nil {
+		return
+	}
+
+	return e.write(bytes)
+}
+
+// TODO encode length with variable-sized encoding?
+//      e.g. first byte starting with `0` is the last byte in the length
+//      will usually save 3 bytes. the question is if it saves or costs encode and/or decode time
+
+// EncodeLength encodes a non-negative length as a uint32.
+// It uses 4 bytes.
+func (e *SemaEncoder) EncodeLength(length int) (err error) {
+	if length < 0 { // TODO is this safety check useful?
+		return fmt.Errorf("cannot encode length below zero: %d", length)
+	}
+
+	l := uint32(length)
+
+	return binary.Write(&e.w, binary.BigEndian, l)
+}
+
+func (e *SemaEncoder) EncodeAddress(address common.Address) (err error) {
+	return e.write(address[:])
+}
+
+func (e *SemaEncoder) write(b []byte) (err error) {
+	_, err = e.w.Write(b)
+	return
+}
+
+func EncodeArray[T any](e *SemaEncoder, arr []T, encodeFn func(T) error) (err error) {
+	// TODO save a bit in the array length for nil check?
+	err = e.EncodeBool(arr == nil)
+	if arr == nil || err != nil {
+		return
+	}
+
+	err = e.EncodeLength(len(arr))
+	if err != nil {
+		return
+	}
+
+	for _, element := range arr {
+		// TODO does this need to include pointer logic for recursive types in arrays to be handled correctly?
+		err = encodeFn(element)
+		if err != nil {
+			return
+		}
+	}
+
+	return
+}
+
+// EncodeMap serializes a map from TypeID to sema.Type.
+func EncodeMap[V sema.Type](e *SemaEncoder, m map[common.TypeID]V, encodeFn func(V) error) (err error) {
+	// ASSUMPTION: map is never `nil`. WHY: EncodeMap only used for Elaboration, which is always fully instantiated.
+
+	err = e.EncodeLength(len(m))
+	if err != nil {
+		return
+	}
+
+	// The order of encoded key-value pairs does not matter so long as we don't rely on hashing encoded Elaborations.
+	for k, v := range m { //nolint:maprangecheck
+		err = e.EncodeString(string(k))
+		if err != nil {
+			return
+		}
+
+		if bufferOffset, usePointer := e.typeDefs[v]; usePointer {
+			err = e.EncodePointer(bufferOffset)
+			if err != nil {
+				return
+			}
+			continue
+		}
+		e.typeDefs[v] = e.w.length
+
+		err = encodeFn(v)
+		if err != nil {
+			return
+		}
+	}
+
+	return
+}
+
+type LengthyWriter struct {
```

Comment:
Maybe don't export this

Reply:
In my next PR, this will be in a library package, `common_codec`, since it's useful for the cadence.Value codec as well.